File: 01/main.exs
# Benchmarks
#
# Name             ips        average  deviation         median         99th %
# part_1        773.23        1.29 ms   ±136.33%        1.20 ms        2.12 ms
# part_2        868.88        1.15 ms   ±150.91%        1.07 ms        1.71 ms
#
# Comparison:
# part_1        773.23 - 1.12x slower +0.142 ms
# part_2        868.88
#
# Memory usage statistics:
#
# Name      Memory usage
# part_1         1.09 MB - 1.56x memory usage +0.39 MB
# part_2         0.70 MB

defmodule Main do
  @moduledoc false

  @spec line_stream() :: Stream.t()
  def line_stream() do
    File.stream!("input")
    |> Stream.map(&String.trim_trailing(&1, "\n"))
  end

  @spec parse_line(String.t()) :: {number(), number()}
  def parse_line(line) when is_binary(line) do
    line
    |> String.split(" ", trim: true)
    |> Enum.map(&String.to_integer/1)
    |> then(fn [left, right] ->
      {left, right}
    end)
  end

  @spec lists() :: {list(), list()}
  def lists() do
    line_stream()
    |> Enum.to_list()
    |> Enum.map(&parse_line/1)
    |> Enum.unzip()
  end

  @spec part_1() :: any()
  def part_1() do
    {left, right} = lists()

    [
      Enum.sort(left),
      Enum.sort(right)
    ]
    |> Enum.zip()
    |> Enum.map(fn {left, right} ->
      abs(right - left)
    end)
    |> Enum.sum()
  end

  @spec part_2() :: any()
  def part_2() do
    {left, right} = lists()
    frequencies = Enum.frequencies(right)

    left
    |> Enum.map(& &1 * Map.get(frequencies, &1, 0))
    |> Enum.sum()
  end
end

IO.puts("Part 1: #{Main.part_1()}")
IO.puts("Part 2: #{Main.part_2()}")

File: 01/main.zig
const std = @import("std");
const FileLineIterator = @import("FileLineIterator.zig");

const input_file = "input";

pub fn main() !void {
    var gpa: std.heap.DebugAllocator(.{}) = .init;
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("Part 1: {d}\n", .{try part1(allocator)});
    std.debug.print("Part 2: {d}\n", .{try part2(allocator)});
}

fn part1(allocator: std.mem.Allocator) !usize {
    var lists = try readLists(allocator);
    defer lists.left.deinit(allocator);
    defer lists.right.deinit(allocator);

    std.mem.sort(isize, lists.left.items, {}, std.sort.asc(isize));
    std.mem.sort(isize, lists.right.items, {}, std.sort.asc(isize));

    var sum: usize = 0;
    for (lists.left.items, lists.right.items) |left, right| {
        sum += @abs(left - right);
    }

    return sum;
}

fn part2(allocator: std.mem.Allocator) !isize {
    var lists = try readLists(allocator);
    defer lists.left.deinit(allocator);
    defer lists.right.deinit(allocator);

    var frequencies: std.AutoHashMap(isize, isize) = .init(allocator);
    defer frequencies.deinit();

    for (lists.right.items) |item| {
        const result = try frequencies.getOrPut(item);
        if (result.found_existing) {
            result.value_ptr.* += 1;
        } else {
            result.value_ptr.* = 1;
        }
    }

    var sum: isize = 0;
    for (lists.left.items) |item| {
        if (frequencies.get(item)) |frequency| {
            sum += item * frequency;
        }
    }

    return sum;
}

fn readLists(allocator: std.mem.Allocator) !struct { left: std.ArrayList(isize), right: std.ArrayList(isize) } {
    var file_line_iterator = try FileLineIterator.init(allocator, input_file);
    defer file_line_iterator.deinit();

    var left_list: std.ArrayList(isize) = .empty;
    var right_list: std.ArrayList(isize) = .empty;

    var line_number: usize = 0;

    while (try file_line_iterator.next()) |line| {
        line_number += 1;
        const parsed = parseLine(line) catch |err| {
            std.debug.print("Line {d}: \"{s}\"\n", .{ line_number, line });
            return err;
        };
        try left_list.append(allocator, parsed.left);
        try right_list.append(allocator, parsed.right);
    }

    return .{
        .left = left_list,
        .right = right_list,
    };
}

fn parseLine(line: []const u8) !struct { left: isize, right: isize } {
    var token_iterator = std.mem.tokenizeScalar(u8, line, ' ');

    const left_token = token_iterator.next() orelse return error.NoLeftToken;
    const right_token = token_iterator.next() orelse return error.NoRightToken;
    if (token_iterator.peek() != null) return error.UnexpectedToken;

    const left = std.fmt.parseInt(isize, left_token, 10) catch return error.NotInteger;
    const right = std.fmt.parseInt(isize, right_token, 10) catch return error.NotInteger;

    return .{
        .left = left,
        .right = right,
    };
}
